web/packages/teleterm/src/ui/DocumentCluster/UnifiedResources.tsx

```diff
@@ -0,0 +1,287 @@
+import { routing } from 'teleterm/ui/uri';
+
+import {
+  ConnectServerActionButton,
+  ConnectKubeActionButton,
+  ConnectDatabaseActionButton,
+} from './actionButtons';
+import { useResourcesContext } from './resourcesContext';
+
+interface UnifiedResourcesProps {
+  clusterUri: uri.ClusterUri;
+}
+
+export function UnifiedResources(props: UnifiedResourcesProps) {
+  const appContext = useAppContext();
+  const { onResourcesRefreshRequest } = useResourcesContext();
+
+  const [params, setParams] = useState<UnifiedResourcesQueryParams>({
+    sort: { fieldName: 'name', dir: 'ASC' },
+  });
+
+  const { documentsService, rootClusterUri } = useWorkspaceContext();
+  const loggedInUser = useWorkspaceLoggedInUser();
+  const { canUse: hasPermissionsForConnectMyComputer, agentCompatibility } =
+    useConnectMyComputerContext();
+
+  const isRootCluster = props.clusterUri === rootClusterUri;
+  const canAddResources = isRootCluster && loggedInUser?.acl?.tokens.create;
+
+  const canUseConnectMyComputer =
+    isRootCluster &&
+    hasPermissionsForConnectMyComputer &&
+    agentCompatibility === 'compatible';
+
+  const { fetch, resources, attempt } = useUnifiedResourcesFetch({
+    fetchFunc: useCallback(
+      async (paginationParams, signal) => {
+        const response = await retryWithRelogin(
+          appContext,
+          props.clusterUri,
+          () =>
+            appContext.resourcesService.listUnifiedResources(
+              {
+                clusterUri: props.clusterUri,
+                searchAsRoles: false,
+                sortBy: {
+                  isDesc: params.sort.dir === 'DESC',
+                  field: params.sort.fieldName,
+                },
+                search: params.search,
+                kindsList: params.kinds,
+                query: params.query,
+                pinnedOnly: params.pinnedOnly,
+                startKey: paginationParams.startKey,
+                limit: paginationParams.limit,
+              },
+              signal
+            )
+        );
+
+        return {
+          startKey: response.nextKey,
+          agents: response.resources,
+          totalCount: response.resources.length,
+        };
+      },
+      [
+        appContext,
+        params.kinds,
+        params.pinnedOnly,
+        params.query,
+        params.search,
+        params.sort.dir,
+        params.sort.fieldName,
+        props.clusterUri,
+      ]
+    ),
+  });
+
+  useEffect(() => {
+    const { cleanup } = onResourcesRefreshRequest(() =>
+      fetch({
+        force: true,
+        fromStart: true,
+      })
+    );
+    return cleanup;
+  }, [onResourcesRefreshRequest, fetch]);
+
+  return (
+    <SharedUnifiedResources
+      params={params}
+      setParams={setParams}
+      updateUnifiedResourcesPreferences={() => alert('Not implemented')}
+      onLabelClick={() => alert('Not implemented')}
+      pinning={{ kind: 'hidden' }}
+      resources={resources
+        .map(resource => {
```

Comment:
If we decide to stick with this function, I think it's best if we move it to a top level definition – it doesn't depend on anything defined within `UnifiedResources` and will make it easier to read the body of this component.